pulsar-client-cpp/lib/HTTPWrapper.cc

```diff
@@ -0,0 +1,241 @@
+                response_.retCode = Response::ResponseFailure;
+                callback_(shared_from_this());
+                return;
+            }
+            // Read the response headers, which are terminated by a blank line.
+            boost::asio::async_read_until(*socketPtr_, *responseStreamPtr_, "\r\n\r\n",
+                                          boost::bind(&HTTPWrapper::handle_read_headers, shared_from_this(),
+                                                      boost::asio::placeholders::error));
+        } else {
+            response_.errCode = err;
+            response_.retCode = Response::ResponseFailure;
+            callback_(shared_from_this());
+        }
+    }
+
+    void HTTPWrapper::handle_read_headers(const boost::system::error_code &err) {
+        // boost::asio::error::eof not handles since HTTP format don't expect EOF before \r\n\r\n
+        if (!err) {
+            std::istream inputStream(responseStreamPtr_.get());
+            std::string header;
+            // response_.headers guaranteed to have atleast one string since reserve called
+            while (std::getline(inputStream, header) && header != "\r") {
+                removeCarriage(header);
+                response_.headers.push_back(header);
+            }
+
+            if (responseStreamPtr_.get()->size() > 0) {
+                // if content doesn't end with \r\n getline doesn't the extract content - remaining characters to be ignored
+                // http://www.boost.org/doc/libs/1_55_0/doc/html/boost_asio/reference/async_read_until/overload1.html
+            }
+
+            // Start reading remaining data until EOF.
+            boost::asio::async_read(*socketPtr_, *responseStreamPtr_,
+                                    boost::asio::transfer_at_least(1),
+                                    boost::bind(&HTTPWrapper::handle_read_content, shared_from_this(),
+                                                boost::asio::placeholders::error));
+        } else {
+            response_.errCode = err;
+            response_.retCode = Response::ResponseFailure;
+            callback_(shared_from_this());
+        }
+    }
+
+    void HTTPWrapper::handle_read_content(const boost::system::error_code &err) {
+        if (!err) {
+            std::istream inputStream(responseStreamPtr_.get());
+            inputStream >> response_.content;
+            boost::asio::async_read(*socketPtr_, *responseStreamPtr_,
+                                    boost::asio::transfer_at_least(1),
+                                    boost::bind(&HTTPWrapper::handle_read_content, shared_from_this(),
+                                                boost::asio::placeholders::error));
+        } else if (err == boost::asio::error::eof) {
+            // eof occurs but responseStreamPtr_ not necessarily emptys
+            LOG_DEBUG("EOF occured");
+            std::istream inputStream(responseStreamPtr_.get());
+            inputStream >> response_.content;
```

Comment:
is the intent to read all of the stream content and store into response_.content?  this only reads the first whitespace delimited string.  try this:

```
std::ostringstream os;
os << inputStream.rdbuf();
response_.content = os.str();
```